docs/rfcs/1005-realm-key-rotation.md

```diff
@@ -0,0 +1,821 @@
+<!-- Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS -->
+
+# No-reencryption Realm key rotation
+
+## 0 Abstract
+
+This RFC proposes a way to rotate realm keys without requiring re-encryption of all data.
+
+## 1 Background
+
+The current realm key rotation system works by replacing the realm key, then re-encrypting
+all realm's vlobs with the new key.
+
+The rational for this approach was that the client only have to deal with a single key at any given time.
+However this has some downsides:
+
+- Reencrypting all vlobs takes some time (though not that much in practice given it's only
+  metadata, the empirical rule being 1000 times less metadata than actual data in Parsec).
+- This concept must be exposed to the user so that he can decide when to do the reencryption
+  maintenance procedure. This is a concept complex to explain so not great for UX :/
+- A realm encryption revision index is kept server-side and send to the client to ensure
+  it is not using the wrong key to decrypt
+- A re-encryption API is needed (the `realm_(start|stop)_reencryption_maintenance` &
+  `vlob_maintenance_(get|save)_reencryption_batch` commands)
+- The concept of maintenance has been introduced: a realm under maintenance cannot be read/write
+- Due to the maintenance, a reencryption must finish no matter what. So all realm owner have to
+  watch for ongoing maintenance an participate in it (even if they are not the initiator of it),
+  just in case the initiator disconnects (to be never seem again) before completion.
+
+On top of that the added security is small (remember, this was a practical choice) given
+the client could have downloaded all metadata when it has access (typically when doing
+file content a search on the workspace mountpoint).
+In addition since Parsec 2.16 the manifests are eagerly downloaded (instead of the
+previous lazy approach), so the reencryption really has no security benefit.
+
+And now the final blow: the maintenance preventing read/write access was such a strong
+limitation that we ended up adding a feature to keep read access possible.
+Now what that means is the client has to in fact deal with up to *two* keys:
+
+1. when a maintenance starts, the client receives the new key
+2. it overwrites the realm key with the new one it the user manifest
+3. now it tries to access some realm data
+   - the maintenance is done, so the new key works. All is good.
+   - the maintenance is still ongoing, so the client has to look into the previous
+     versions of the user manifest to find the old key. This is messy :'(
+
+## 2 General approach
+
+The idea is as follow:
+
+- remove the whole maintenance logic: a realm is always accessible in read&write
+- remove the whole re-encryption logic: realm vlobs are never reencrypted
+- remove the whole message logic: we use a special ad-hoc command to access the
```

Comment:
The "message logic" should be briefly introduced/explained in section 1. I'll continue reading to understand ^^